content/things/webgpu.md

```diff
@@ -0,0 +1,586 @@
+
+The GPU now rasterizes the triangles, meaning the GPU figures out which pixels each triangle covers on the screen. Each pixel is then processed by the _fragment shader_, which has access to the pixel coordinates but also the auxillary data to decide which color that pixel should be. When used correctly, amazing 3D graphics can be created with this process.
+
+This system of passing data to a vertex shader, then to a fragment shader and then outputting it directly onto the screen is called a _pipeline_, and in WebGPU you have to explicitly define your pipeline.
+
+### Pipelines
+
+Currently, WebGPU allows you to create two types of pipelines: A Render Pipeline and a Compute Pipeline. As the name suggest, the Render Pipeline renders something, meaning it creates a 2D image. That image needn’t be on screen, but could just be rendered to memory (which is called a Framebuffer). A Compute Pipeline is more generic in that it returns a buffer, which can contain any sort of data. For the remainder of this blog post I’ll focus on Compute Pipelines, as I like to think of Render Pipelines as a specialization/optimization of Compute Pipelines. Now, this is not only historically backwards — the compute pipeline was built as an abstraction over the very purpose-built rendering pipeline — but actually considerably understates that these pipelines are physically different circuits in your GPU. In terms of the API, however, I find my mental model quite good. <mark>In the future, it seems likely that more types of pipelines — maybe a Raytracing Pipeline — get added to WebGPU.</mark>
+
+With WebGPU, a pipeline consists of one (or more) programmable stages, where each stage is defined by a shader and an entry point. A Compute Pipline has a single `compute` stage, while a Render Pipeline would have a `vertex` and a `fragment` stage:
+
+```js
+const module = device.createShaderModule({
+  code: `
+    @stage(compute) @workgroup_size(64)
+    fn main() {
+      // Pointless!
+    }
+  `,
+});
+
+const pipeline = device.createComputePipeline({
+  compute: {
+    module,
+    entryPoint: "main",
+  },
+});
+```
+
+<mark>Should I recommend using `createComputePipelineAsync` instead?</mark>
+
+This is the first time [WGSL] (pronounced “wig-sal”), the WebGPU Shading Language, makes an appearance. WGSL feels like a cross-over of Rust and GLSL to me. It’s Rust syntax with GLSL’ global functions (like `dot()`, `norm()`, `len()`, ...), types (like `vec2`, `mat4x4`, ...) and the swizzling notation (like `some_vec.xxy`, ...). The browser <mark>will most likely is compile your WGSL to [SPIR-V], which is a binary intermediate format standardized by the Khronos Group. You can think of SPIR-V as the LLVM of parallel programming language compilers, but you don’t need to learn it as you probably won’t come into direct contact with SPIR-V.</mark>
```

Comment:
At this point it seems like the browser will only compile to SPIR-V for systems that use a Vulkan backend. It would be more accurate to say that the browser will cross-compile the WGSL code into whatever the native shader language is for the underlying API (HLSL for D3D12, MSL for Metal, SPIR-V for Vulkan) and then let that API's native toolchain do the rest of the work. The background on SPIR-V here is nice, though! Maybe just highlight it as one possible route instead the most likely one.